res/res_agi.c

```diff
@@ -1026,6 +1026,103 @@ static struct agi_cmd *get_agi_cmd(struct ast_channel *chan)
+static int add_silence(struct ast_channel *chan, struct ast_frame *f, struct ast_filestream *fs, int stream_no)
+{
+	int last_seq=0, f_no=0, j=0;
+	long int ts_diff=0, f_ptime;
+	short buf[f->samples];
+        struct ast_frame *duped_frame = NULL;
+
+	unsigned char g729_filler[] = {
+                114, 170, 255, 103, 54, 82, 216, 110, 255, 81,
+                114, 170, 255, 103, 54, 82, 216, 110, 255, 81,
+                114, 170, 255, 103, 54, 82, 216, 110, 255, 81,
+                114, 170, 255, 103, 54, 82, 216, 110, 255, 81,
+                114, 170, 255, 103, 54, 82, 216, 110, 255, 81,
+                114, 170, 255, 103, 54, 82, 216, 110, 255, 81,
+                114, 170, 255, 103, 54, 82, 216, 110, 255, 81,
+                114, 170, 255, 103, 54, 82, 216, 110, 255, 81,
+                114, 170, 255, 103, 54, 82, 216, 110, 255, 81,
+                114, 170, 255, 103, 54, 82, 216, 110, 255, 81,
+                114, 170, 255, 103, 54, 82, 216, 110, 255, 81,
+                114, 170, 255, 103, 54, 82, 216, 110, 255, 81,
+                114, 170, 255, 103, 54, 82, 216, 110, 255, 81,
+                114, 170, 255, 103, 54, 82, 216, 110, 255, 81
+        }; /*! BEST SO FAR - 2 */
+
+	if (stream_no == 1) {
+		f_ptime=ast_channel_get_s1_ptime(chan);
+		if(f_ptime < 5)
+			f_ptime=20;
+
+        	ts_diff = f->ts - ast_channel_get_s1_last_ts(chan);
+        	last_seq = ast_channel_get_s1_last_seq(chan);
+		f_no=ast_channel_get_s1_last_ts(chan)+f_ptime;
+
+		/*! Save the ts and sequence number for the next check */
+        	ast_channel_set_s1_last_ts(chan, f->ts);
+        	ast_channel_set_s1_last_seq(chan, f->seqno);
+	} else {
+		f_ptime = ast_channel_get_s2_ptime(chan);
+		if(f_ptime < 5)
+                        f_ptime=20;
+		
+		ts_diff = f->ts - ast_channel_get_s2_last_ts(chan);
+		last_seq = ast_channel_get_s2_last_seq(chan);
+		f_no=ast_channel_get_s2_last_ts(chan)+f_ptime;
+
+		/*! Save the ts and sequence number for the next check */
+                ast_channel_set_s2_last_ts(chan, f->ts);
+                ast_channel_set_s2_last_seq(chan, f->seqno);
+	}
+
+        ast_debug(3, "STREAM %d -- len: %ld samples: %d datalength: %d seqno: %d timestamp: %0.4f ts: %ld\n", stream_no, f->len, f->samples, f->datalen, f->seqno, (float)f->ts/1000.00, f->ts);
+
+        if (ts_diff >= (2*f_ptime)) { // Twice the ptime size
+        	if ((f->seqno - last_seq) > 1)
+                	ast_log(LOG_WARNING, "STREAM %d -- GAP: %f\t No of Frames Lost: %d\n", stream_no, (float)ts_diff/1000.0, f->seqno - last_seq - 1);
+                else
+                	ast_log(LOG_WARNING, "STREAM %d -- Frame (%d) receieved after %0.3f sec\n", stream_no, f->seqno, (float)ts_diff/1000.0);
+
+                /* Lets try generating silent frame */
+                duped_frame = ast_frdup(f);
+
+		switch (f->subclass.format.id) {
+                case AST_FORMAT_G729A:
+                	memcpy(buf, g729_filler, f->len);
```

Comment:
f->len provides the number of bytes to be copied into the buf from g729_filler.

Reply:
Also the size of the buf should be equal to f->samples and is defined because for 
1. PCMA & PCMU the samples = f->datalen
2. For G729 it is samples = f->datalen * 8